## Preprocessing

In [21]:
# Import our dependencies
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

# Import pandas and read the charity_data.csv from the provided cloud URL.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [22]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df = application_df.drop(columns = ['EIN', 'NAME'])


In [23]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [24]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64

In [25]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
cutoff_value = 500
app_count = application_df['APPLICATION_TYPE'].value_counts()
application_types_to_replace = app_count[app_count < cutoff_value].index.tolist()

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [26]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_list = application_df['CLASSIFICATION'].value_counts()
classification_list

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C1248        1
C6100        1
C1820        1
C1900        1
C2150        1
Name: count, Length: 71, dtype: int64

In [27]:
# You may find it helpful to look at CLASSIFICATION value counts >1
classification_list[classification_list>1]


CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C6000       15
C1800       15
C1250       14
C8200       11
C1238       10
C1278       10
C1237        9
C1235        9
C7210        7
C1720        6
C2400        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1256        2
C1246        2
C1234        2
C1267        2
Name: count, dtype: int64

In [28]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
cutoff_value = 1000
class_count = application_df['CLASSIFICATION'].value_counts()
classifications_to_replace = class_count[class_count < cutoff_value].index.tolist()

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [29]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df, dtype=int)
application_df.head()


STATUS  ASK_AMT  IS_SUCCESSFUL  APPLICATION_TYPE_Other  \
0       1     5000              1                       0   
1       1   108590              1                       0   
2       1     5000              0                       0   
3       1     6692              1                       0   
4       1   142590              1                       0   

   APPLICATION_TYPE_T10  APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  \
0                     1                     0                    0   
1                     0                     0                    1   
2                     0                     0                    0   
3                     0                     0                    1   
4                     0                     0                    1   

   APPLICATION_TYPE_T4  APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  ...  \
0                    0                    0                    0  ...   
1                    0                    0                    0  ...   
2                    0                    1                    0  ...   
3                    0                    0                    0  ...   
4                    0                    0                    0  ...   

   INCOME_AMT_1-9999  INCOME_AMT_10000-24999  INCOME_AMT_100000-499999  \
0                  0                       0                         0   
1                  1                       0                         0   
2                  0                       0                         0   
3                  0                       1                         0   
4                  0                       0                         1   

   INCOME_AMT_10M-50M  INCOME_AMT_1M-5M  INCOME_AMT_25000-99999  \
0                   0                 0                       0   
1                   0                 0                       0   
2                   0                 0                       0   
3                   0                 0                       0   
4                   0                 0                       0   

   INCOME_AMT_50M+  INCOME_AMT_5M-10M  SPECIAL_CONSIDERATIONS_N  \
0                0                  0                         1   
1                0                  0                         1   
2                0                  0                         1   
3                0                  0                         1   
4                0                  0                         1   

   SPECIAL_CONSIDERATIONS_Y  
0                         0  
1                         0  
2                         0  
3                         0  
4                         0  

[5 rows x 44 columns]

In [30]:
# Split our preprocessed data into our features and target arrays
X = application_df.drop('IS_SUCCESSFUL', axis =1).values
y = application_df['IS_SUCCESSFUL'].values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [31]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Compile, Train and Evaluate the Model

In [32]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(tf.keras.layers.Dense(units=128, activation='relu', input_dim=X_train.shape[1]))

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=64, activation='relu'))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

# Check the structure of the model
nn.summary()

/usr/local/lib/python3.11/dist-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)


Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_3 (Dense)                 │ (None, 128)            │         5,632 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_4 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_5 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 13,953 (54.50 KB)

 Trainable params: 13,953 (54.50 KB)

 Non-trainable params: 0 (0.00 B)

In [33]:
# Compile the model
nn.compile(loss='binary_crossentropy', optimizer='adam', metrics=['accuracy'])

In [34]:
# Train the model
history = nn.fit(X_train, y_train, epochs=100, batch_size=64, validation_data=(X_test, y_test))

Epoch 1/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.4887 - loss: 29469.7539 - val_accuracy: 0.4668 - val_loss: 3551.3560
Epoch 2/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.4721 - loss: 23903.0801 - val_accuracy: 0.5332 - val_loss: 251186.2188
Epoch 3/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.5047 - loss: 57239.2695 - val_accuracy: 0.5577 - val_loss: 170.4359
Epoch 4/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.4984 - loss: 16148.6465 - val_accuracy: 0.6273 - val_loss: 470.8669
Epoch 5/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 2s 4ms/step - accuracy: 0.4985 - loss: 15844.4883 - val_accuracy: 0.5332 - val_loss: 133369.6406
Epoch 6/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 2s 3ms/step - accuracy: 0.5128 - loss: 53506.9219 - val_accuracy: 0.4668 - val_loss: 1424.0593
Epoch 7/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 1s 3ms/step - accuracy: 0.5058 - loss: 12524.5410 - val_accuracy: 0.4668 - val_loss: 2047.7665
Epoch 8/100
402/402 ━━━━━━━━━━━━━━━━━━━━ 1s 3m

In [35]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 1s - 2ms/step - accuracy: 0.6612 - loss: 0.7834
Loss: 0.7833737134933472, Accuracy: 0.6612244844436646


In [36]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity.h5')

## Optimize the Model

In [38]:
# Create the optimized model with more neurons
nn_optimized = tf.keras.models.Sequential()

# First hidden layer with more neurons
nn_optimized.add(tf.keras.layers.Dense(units=256, activation='relu', input_dim=X_train.shape[1]))
nn_optimized.add(tf.keras.layers.Dropout(0.2))  # Dropout to avoid overfitting

# Second hidden layer with more neurons
nn_optimized.add(tf.keras.layers.Dense(units=128, activation='relu'))
nn_optimized.add(tf.keras.layers.Dropout(0.2))  # Dropout to avoid overfitting

# Third hidden layer (to make it deeper)
nn_optimized.add(tf.keras.layers.Dense(units=64, activation='relu'))

# Output layer
nn_optimized.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))  # Sigmoid for binary classification

# Check the summary of the model to confirm structure
nn_optimized.summary()


Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_6 (Dense)                 │ (None, 256)            │        11,264 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout (Dropout)               │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_7 (Dense)                 │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_1 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 52,481 (205.00 KB)

 Trainable params: 52,481 (205.00 KB)

 Non-trainable params: 0 (0.00 B)

In [39]:
# Define the model using the 'tanh' activation function
nn_optimized = tf.keras.models.Sequential()

# First hidden layer with tanh activation
nn_optimized.add(tf.keras.layers.Dense(units=256, input_dim=X_train.shape[1]))
nn_optimized.add(tf.keras.layers.Activation('tanh'))  # 'tanh' activation instead of ReLU
nn_optimized.add(tf.keras.layers.Dropout(0.2))  # Dropout to avoid overfitting

# Second hidden layer with tanh activation
nn_optimized.add(tf.keras.layers.Dense(units=128))
nn_optimized.add(tf.keras.layers.Activation('tanh'))  # 'tanh' activation
nn_optimized.add(tf.keras.layers.Dropout(0.2))  # Dropout to avoid overfitting

# Third hidden layer
nn_optimized.add(tf.keras.layers.Dense(units=64))
nn_optimized.add(tf.keras.layers.Activation('tanh'))  # 'tanh' activation

# Output layer
nn_optimized.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))  # Sigmoid for binary classification

# Check the model summary
nn_optimized.summary()


Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_10 (Dense)                │ (None, 256)            │        11,264 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation (Activation)         │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_2 (Dropout)             │ (None, 256)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 128)            │        32,896 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation_1 (Activation)       │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dropout_3 (Dropout)             │ (None, 128)            │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 64)             │         8,256 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ activation_2 (Activation)       │ (None, 64)             │             0 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_13 (Dense)                │ (None, 1)              │            65 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 52,481 (205.00 KB)

 Trainable params: 52,481 (205.00 KB)

 Non-trainable params: 0 (0.00 B)

* Reduce the Epochs to 100:

In [40]:
# Compile the optimized model (same as before)
nn_optimized.compile(optimizer='adam',
                     loss='binary_crossentropy',
                     metrics=['accuracy'])

# Train the model with reduced epochs (100 epochs)
history_optimized_100 = nn_optimized.fit(X_train, y_train, epochs=100, batch_size=32, validation_data=(X_test, y_test))

# Check the training and testing accuracy
print(f"Optimized Training Accuracy (100 epochs): {history_optimized_100.history['accuracy'][-1]}")
print(f"Optimized Testing Accuracy (100 epochs): {history_optimized_100.history['val_accuracy'][-1]}")


Epoch 1/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.5160 - loss: 0.7058 - val_accuracy: 0.5332 - val_loss: 0.6967
Epoch 2/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.5281 - loss: 0.6923 - val_accuracy: 0.5332 - val_loss: 0.6916
Epoch 3/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5320 - loss: 0.6913 - val_accuracy: 0.5332 - val_loss: 0.6923
Epoch 4/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 7s 5ms/step - accuracy: 0.5280 - loss: 0.6923 - val_accuracy: 0.4668 - val_loss: 0.7027
Epoch 5/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.5225 - loss: 0.6926 - val_accuracy: 0.5332 - val_loss: 0.6909
Epoch 6/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.5230 - loss: 0.6926 - val_accuracy: 0.5332 - val_loss: 0.6910
Epoch 7/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.5299 - loss: 0.6920 - val_accuracy: 0.5332 - val_loss: 0.6931
Epoch 8/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 4ms/step - accuracy: 0.5235 - loss: 0.6921 - val_accu

* Increase the Epochs to 300:

In [41]:
# Compile the optimized model (same as before)
nn_optimized.compile(optimizer='adam',
                     loss='binary_crossentropy',
                     metrics=['accuracy'])

# Train the model with increased epochs (300 epochs)
history_optimized_300 = nn_optimized.fit(X_train, y_train, epochs=300, batch_size=32, validation_data=(X_test, y_test))

# Check the training and testing accuracy
print(f"Optimized Training Accuracy (300 epochs): {history_optimized_300.history['accuracy'][-1]}")
print(f"Optimized Testing Accuracy (300 epochs): {history_optimized_300.history['val_accuracy'][-1]}")


Epoch 1/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.5249 - loss: 0.6924 - val_accuracy: 0.5332 - val_loss: 0.6913
Epoch 2/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 3ms/step - accuracy: 0.5190 - loss: 0.6928 - val_accuracy: 0.5332 - val_loss: 0.6915
Epoch 3/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 6s 4ms/step - accuracy: 0.5232 - loss: 0.6924 - val_accuracy: 0.5332 - val_loss: 0.6918
Epoch 4/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.5269 - loss: 0.6920 - val_accuracy: 0.5332 - val_loss: 0.6910
Epoch 5/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 4s 5ms/step - accuracy: 0.5177 - loss: 0.6929 - val_accuracy: 0.5332 - val_loss: 0.6917
Epoch 6/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 4ms/step - accuracy: 0.5306 - loss: 0.6919 - val_accuracy: 0.5332 - val_loss: 0.6917
Epoch 7/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 3s 3ms/step - accuracy: 0.5172 - loss: 0.6929 - val_accuracy: 0.4668 - val_loss: 0.6946
Epoch 8/300
804/804 ━━━━━━━━━━━━━━━━━━━━ 5s 3ms/step - accuracy: 0.5149 - loss: 0.6933 - val_accu

In [42]:
# Evaluate the optimized model using the test data
model_loss, model_accuracy = nn_optimized.evaluate(X_test, y_test, verbose=2)
print(f"Optimized Model Loss: {model_loss}, Optimized Model Accuracy: {model_accuracy}")


268/268 - 1s - 4ms/step - accuracy: 0.5332 - loss: 0.6920
Optimized Model Loss: 0.6920185089111328, Optimized Model Accuracy: 0.5331778526306152


In [44]:
# save the file
nn_optimized.save('AlphabetSoupCharity_Optimization.h5')


### **Optimized Model Evaluation**

**Test Set Performance:**
- **Accuracy:** 53.32%
- **Loss:** 0.6920

**Interpretation:**
- **Accuracy:** The model correctly predicted the success of approximately 53.32% of the test samples. While this is a step forward, it falls short of the desired 75% accuracy target.
- **Loss:** A loss value of 0.6920 indicates that the model's predictions are still somewhat off from the actual outcomes. Lower loss values typically signify better model performance.

**Next Steps for Improvement:**
To enhance the model's performance, consider the following strategies:
- **Hyperparameter Tuning:** Experiment with different learning rates, batch sizes, and numbers of epochs.
- **Feature Engineering:** Explore additional features or transformations that might better capture the underlying patterns.
- **Model Architecture Adjustments:** Test with more complex architectures, such as adding more layers or using different activation functions.
- **Regularization Techniques:** Implement methods like dropout or L2 regularization to prevent overfitting.
